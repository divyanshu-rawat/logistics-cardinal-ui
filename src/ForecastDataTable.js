import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import { TableHeaderColumn } from 'react-bootstrap-table';
import styled from 'styled-components';

import Table from 'components/Table';
import StyledRadioInput from './ForecastStyledRadioInput';

const RelativeDiv = styled('div')`
  position: relative;
  padding: 8px;
  display: flex;
  line-height: 22px;
  justify-content: space-between;
  span {
    padding: 5px;
  }
  &.selected {
    border: 1px solid #3f9bfc;
    padding: 7px;
  }
  .percents {
    color: ${(props) => props.theme.inactive};
  }
`;

const RelativeDivForecasts = styled(RelativeDiv)`
  background-color: ${(props) => props.theme.greyDeep};
`;

const RelativeDivActuals = styled(RelativeDiv)`
  background-color: ${(props) => props.theme.greyMedium};
`;

const BarDivActuals = styled('div')`
  position: absolute;
  height: 30px;
  background-color: rgba(150, 162, 176, 0.2);
`;

const BarDivForecasts = styled(BarDivActuals)`
  background-color: rgba(63, 155, 252, 0.3);
`;

const StyledTable = styled(Table)`
  .table > thead > tr > th,
  .table > tbody > tr > td {
    background-color: ${(props) => props.theme.greyMedium};
    padding: 0;
    outline: 0;
    border-right: 1px solid ${(props) => props.theme.greyLight};
    &:last-child {
      border: 0;
    }
    &:first-child {
      border-right-width: 2px;
    }
    &:focus,
    &:active {
      outline: 0;
    }
  }
  .table > thead > tr > th {
    padding-left: 8px;
    input {
      margin-top: 13px;
    }
  }
  .react-bs-container-body .table {
    margin-bottom: 20px;
  }
  .table > tbody > tr {
    border-bottom: 1px solid ${(props) => props.theme.greyLight};
  }
  .table > thead > tr {
    border-bottom: 2px solid ${(props) => props.theme.greyLight};
  }
  .react-bs-container-body .table {
    margin-bottom: 25px;
  }
`;

const WeekDiv = styled('div')`
  padding: 8px;
  > div {
    line-height: 30px;
  }
`;

export const cellFormatter = (dataObj, onClickHandler, selectedDays = []) => {
  const { biggestOrdersNumber, week, cell, day } = dataObj;
  const percent = cell.value / biggestOrdersNumber * 100;
  const today = moment().startOf('day');
  const weekYearArray = week.match(/\d+/g);
  const year = `${today.format('YYYY').substr(0, 2)}${weekYearArray[1]}`;
  const isoWeek = parseInt(weekYearArray[0], 10);
  const date = today
    .clone()
    .year(year)
    .isoWeek(isoWeek)
    .startOf('isoWeek');

  if (day) {
    date.day(parseInt(day, 10)).startOf('day');
  } else {
    date.endOf('isoWeek');
  }

  let BarComponent = BarDivActuals;
  let CellComponent = RelativeDivActuals;

  if (date.isSameOrAfter(today)) {
    BarComponent = BarDivForecasts;
    CellComponent = RelativeDivForecasts;
  }

  const cellProps = {};
  if (day) {
    const additionalData = { day, week };
    cellProps.onClick = () =>
      onClickHandler({
        dataObj: { ...cell, ...additionalData },
        date: moment()
          .year(year)
          .isoWeek(isoWeek)
          .day(parseInt(day, 10))
          .startOf('day'),
      });
  }

  selectedDays.some((adj) => {
    if (adj.dataObj.day === day && adj.dataObj.week === week) {
      cellProps.className = 'selected';
      return true;
    }
    // eslint-disable-line no-else-return
    cellProps.className = '';

    return false;
  });

  return (
    <CellComponent {...cellProps}>
      <BarComponent style={{ width: `${percent}%` }} />
      <span>{cell.value}</span>
      <span className="percents">
        {cell.change || cell.change === 0 ? `${Math.round(cell.change)}%` : ''}
      </span>
    </CellComponent>
  );
};

export const weekCellFormatter = (
  cell,
  backgroundColor,
  week,
  selectRowProp,
) => {
  const checked = selectRowProp.selected.indexOf(week) > -1;
  const onChange = (isSelected) => selectRowProp.onSelect(week, isSelected);
  return (
    <WeekDiv>
      <StyledRadioInput
        id={cell}
        label={cell}
        onChange={onChange}
        checked={checked}
        color={backgroundColor}
      />
    </WeekDiv>
  );
};

export function ForecastDataTable(props) {
  const {
    tableValues,
    sortedDays,
    selectRowProp,
    biggestOrdersNumber,
    biggestTotalValue,
    colors,
    onClickHandler,
    selectedDays,
  } = props;

  return (
    <StyledTable data={tableValues} pagination={false}>
      <TableHeaderColumn
        dataAlign="center"
        width="100px"
        isKey
        dataField="week"
        csvHeader="week"
        dataFormat={(cell, row) =>
          weekCellFormatter(cell, colors[row.week], row.week, selectRowProp)
        }
      >
        <StyledRadioInput
          id="allWeeks"
          label="Week"
          color="#6DB324"
          onChange={selectRowProp.onSelectAll}
          checked={selectRowProp.selected.length === tableValues.length}
        />
      </TableHeaderColumn>
      {sortedDays.map((dayNr) => {
        const dayStr = moment()
          .day(dayNr)
          .format('dddd');
        return (
          <TableHeaderColumn
            key={dayNr}
            dataField={dayNr}
            csvHeader={dayStr}
            dataAlign="center"
            dataFormat={(cell, row) =>
              cellFormatter(
                { biggestOrdersNumber, week: row.week, cell, day: dayNr },
                onClickHandler,
                selectedDays,
              )
            }
          >
            {dayStr}
          </TableHeaderColumn>
        );
      })}
      <TableHeaderColumn
        dataAlign="center"
        dataField="total"
        csvHeader="total"
        dataFormat={(cell, row) =>
          cellFormatter({
            biggestOrdersNumber: biggestTotalValue,
            week: row.week,
            cell,
          })
        }
      >
        Total
      </TableHeaderColumn>
    </StyledTable>
  );
}

ForecastDataTable.propTypes = {
  sortedDays: PropTypes.array.isRequired,
  tableValues: PropTypes.array.isRequired,
  selectRowProp: PropTypes.object.isRequired,
  colors: PropTypes.object.isRequired,
  biggestOrdersNumber: PropTypes.number.isRequired,
  biggestTotalValue: PropTypes.number.isRequired,
  onClickHandler: PropTypes.func.isRequired,
  selectedDays: PropTypes.array.isRequired,
};

export default ForecastDataTable;
