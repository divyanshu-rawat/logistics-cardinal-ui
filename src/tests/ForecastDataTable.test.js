import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { noop } from 'lodash';

import ForecastDataTable, { cellFormatter } from '../ForecastDataTable';
import { tableValues, charts } from './dataBasedOnFixtures';

describe('<ForecastDataTable />', () => {
  let renderedComponent;

  beforeEach(() => {
    const sortedDays = Object.keys(charts).sort();
    const selectRowProp = {
      mode: 'checkbox',
      selected: ['52', '1', '2'],
      onSelect: noop,
      onSelectAll: noop,
    };
    renderedComponent = shallow(
      <ForecastDataTable
        tableValues={tableValues}
        selectRowProp={selectRowProp}
        sortedDays={sortedDays}
        biggestOrdersNumber={67}
        biggestTotalValue={295}
        colors={{}}
        onClickHandler={noop}
        selectedDays={[]}
      />,
    );
  });

  it('should render the component', () => {
    const tree = toJson(renderedComponent);
    expect(tree).toMatchSnapshot();
  });

  describe('cellFormatter', () => {
    let renderedCellComponent;
    beforeEach(() => {
      renderedCellComponent = shallow(
        cellFormatter(
          {
            biggestOrdersNumber: 67,
            week: 'W2-18',
            cell: { value: 51, change: -16 },
            day: 1,
          },
          noop,
          [{ dataObj: { day: 1, week: 'W2-18' } }],
        ),
      );
    });

    it('should render the cell component', () => {
      const tree = toJson(renderedCellComponent);
      expect(tree).toMatchSnapshot();
    });
  });
});
