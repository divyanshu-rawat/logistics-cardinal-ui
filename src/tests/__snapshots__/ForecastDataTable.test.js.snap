// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForecastDataTable /> cellFormatter should render the cell component 1`] = `
<styled.div
  className="selected sc-ifAKCX deIPf"
  onClick={[Function]}
>
  <styled.div
    style={
      Object {
        "width": "76.11940298507463%",
      }
    }
  />
  <span>
    51
  </span>
  <span
    className="percents"
  >
    -16%
  </span>
</styled.div>
`;

exports[`<ForecastDataTable /> should render the component 1`] = `
<Styled(Styled(Table))
  data={
    Array [
      Object {
        "1": Object {
          "change": -16,
          "value": 51,
        },
        "2": Object {
          "change": 100,
          "value": 55,
        },
        "3": Object {
          "change": 354,
          "value": 59,
        },
        "4": Object {
          "change": 215,
          "value": 63,
        },
        "5": Object {
          "change": 857,
          "value": 67,
        },
        "6": Object {
          "value": undefined,
        },
        "7": Object {
          "value": undefined,
        },
        "total": Object {
          "change": 54,
          "value": 295,
        },
        "week": "W2-18",
      },
      Object {
        "1": Object {
          "value": 61,
        },
        "2": Object {
          "value": 0,
        },
        "3": Object {
          "value": 13,
        },
        "4": Object {
          "value": 20,
        },
        "5": Object {
          "value": 7,
        },
        "6": Object {
          "change": 4200,
          "value": 43,
        },
        "7": Object {
          "change": 840,
          "value": 47,
        },
        "total": Object {
          "change": 3083,
          "value": 191,
        },
        "week": "W1-18",
      },
      Object {
        "1": Object {
          "value": undefined,
        },
        "2": Object {
          "value": undefined,
        },
        "3": Object {
          "value": undefined,
        },
        "4": Object {
          "value": undefined,
        },
        "5": Object {
          "value": undefined,
        },
        "6": Object {
          "value": 1,
        },
        "7": Object {
          "value": 5,
        },
        "total": Object {
          "value": 6,
        },
        "week": "W52-17",
      },
    ]
  }
  pagination={false}
>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="week"
    dataAlign="center"
    dataField="week"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={true}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width="100px"
  >
    <StyledRadioInput
      checked={true}
      color="#6DB324"
      id="allWeeks"
      label="Week"
      onChange={[Function]}
    />
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Monday"
    dataAlign="center"
    dataField="1"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="1"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Monday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Tuesday"
    dataAlign="center"
    dataField="2"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="2"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Tuesday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Wednesday"
    dataAlign="center"
    dataField="3"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="3"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Wednesday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Thursday"
    dataAlign="center"
    dataField="4"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="4"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Thursday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Friday"
    dataAlign="center"
    dataField="5"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="5"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Friday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Saturday"
    dataAlign="center"
    dataField="6"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="6"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Saturday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="Sunday"
    dataAlign="center"
    dataField="7"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    key="7"
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Sunday
  </TableHeaderColumn>
  <TableHeaderColumn
    className=""
    columnClassName=""
    columnTitle={false}
    csvFieldType="string"
    csvHeader="total"
    dataAlign="center"
    dataField="total"
    dataFormat={[Function]}
    dataSort={false}
    defaultASC={false}
    editColumnClassName=""
    editable={true}
    expandable={true}
    filterFormatted={false}
    headerTitle={true}
    hidden={false}
    hiddenOnInsert={false}
    invalidEditColumnClassName=""
    isKey={false}
    keyValidator={false}
    searchable={true}
    sortIndicator={true}
    width={null}
  >
    Total
  </TableHeaderColumn>
</Styled(Styled(Table))>
`;
